File: windows/src/GoalMaker.Core/Planning/RitualReminder.cs
namespace GoalMaker.Core.Planning;

/// <summary>
/// The evening Plan tomorrow reminder (docs/reminders.md, pinned by the 'ritual' cases in
/// contracts/vectors/reminders.json). It rings once per planning day at the time the owner chose,
/// unless the ritual was already done or skipped that day on either device. Quiet hours don't move it:
/// the owner picked the time.
/// </summary>
public static class RitualReminder
{
    private const int DaysAhead = 3;

    /// <summary>20:00, until the owner picks another time or switches it off.</summary>
    public static TimeOnly DefaultTime { get; } = new(20, 0);

    /// <summary>The first moment of planning day <paramref name="day"/> whose clock reads <paramref name="time"/>: after midnight when the time comes before the day's start.</summary>
    public static DateTime MomentOf(DateOnly day, TimeOnly time, int dayStartHour) =>
        time.Hour >= dayStartHour ? day.ToDateTime(time) : day.AddDays(1).ToDateTime(time);

    /// <summary>The planning day whose reminder to show at <paramref name="now"/>: today's, when its moment came after <paramref name="since"/>. Never an earlier day.</summary>
    public static DateOnly? Due(TimeOnly? time, int dayStartHour, IReadOnlySet<DateOnly> ran, DateTime since, DateTime now)
    {
        if (time is not { } at)
        {
            return null;
        }

        var today = PlanningDay.Of(now, dayStartHour);
        var moment = MomentOf(today, at, dayStartHour);
        return !ran.Contains(today) && moment > since && moment <= now ? today : null;
    }

    /// <summary>The moment the alarm waits for after <paramref name="now"/>, or null when the reminder is off.</summary>
    public static DateTime? Next(TimeOnly? time, int dayStartHour, IReadOnlySet<DateOnly> ran, DateTime now)
    {
        if (time is not { } at)
        {
            return null;
        }

        var day = PlanningDay.Of(now, dayStartHour);
        for (var step = 0; step < DaysAhead; step++, day = day.AddDays(1))
        {
            var moment = MomentOf(day, at, dayStartHour);
            if (!ran.Contains(day) && moment > now)
            {
                return moment;
            }
        }

        return null;
    }

    /// <summary>Whether a reminder on screen for <paramref name="day"/> has to go: the ritual ran, or the planning day moved on.</summary>
    public static bool Stale(DateOnly day, int dayStartHour, IReadOnlySet<DateOnly> ran, DateTime now) =>
        ran.Contains(day) || PlanningDay.Of(now, dayStartHour) != day;
}

File: android/app/src/main/kotlin/com/goalmaker/app/domain/planning/RitualReminder.kt
package com.goalmaker.app.domain.planning

import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

/**
 * The evening Plan tomorrow reminder (docs/reminders.md, pinned by the 'ritual' cases in
 * contracts/vectors/reminders.json). It rings once per planning day at the time the owner chose,
 * unless the ritual was already done or skipped that day on either device. Quiet hours don't move it:
 * the owner picked the time.
 */
object RitualReminder {
    /** 20:00, until the owner picks another time or switches it off. */
    val DEFAULT_TIME: LocalTime = LocalTime.of(20, 0)

    /** The first moment of planning day [day] whose clock reads [time]: after midnight when [time] comes before the day's start. */
    fun momentOf(day: LocalDate, time: LocalTime, dayStartHour: Int): LocalDateTime =
        if (time.hour >= dayStartHour) day.atTime(time) else day.plusDays(1).atTime(time)

    /** The planning day whose reminder to show at [now]: today's, when its moment came after [since]. Never an earlier day. */
    fun due(time: LocalTime?, dayStartHour: Int, ran: Set<LocalDate>, since: LocalDateTime, now: LocalDateTime): LocalDate? {
        if (time == null) return null
        val today = PlanningDay.of(now, dayStartHour)
        val moment = momentOf(today, time, dayStartHour)
        return today.takeIf { it !in ran && moment.isAfter(since) && !moment.isAfter(now) }
    }

    /** The moment the alarm waits for after [now], or null when the reminder is off. */
    fun next(time: LocalTime?, dayStartHour: Int, ran: Set<LocalDate>, now: LocalDateTime): LocalDateTime? {
        if (time == null) return null
        var day = PlanningDay.of(now, dayStartHour)
        repeat(DAYS_AHEAD) {
            val moment = momentOf(day, time, dayStartHour)
            if (day !in ran && moment.isAfter(now)) return moment
            day = day.plusDays(1)
        }
        return null
    }

    /** Whether a reminder on screen for [day] has to go: the ritual ran, or the planning day moved on. */
    fun stale(day: LocalDate, dayStartHour: Int, ran: Set<LocalDate>, now: LocalDateTime): Boolean =
        day in ran || PlanningDay.of(now, dayStartHour) != day

    private const val DAYS_AHEAD = 3
}
